[clang-tidy] introduce a unused local non trival variable check

rockwotj opened the pull request:
Introduce a new (off by default) clang tidy check to ensure that variables of a specific type are always used even if -Wunused-variables wouldn't generate a warning.

This check has already caught a couple of different bugs on the codebase I work on, where not handling a future means that lifetimes may not be kept alive properly as an async chunk of code may run after a class has been destroyed, etc.

I would like to upstream it because I believe there could be other applications of this check that would be useful in different contexts.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -187,6 +187,13 @@ New checks
+  <clang-tidy/checks/misc/must-use>` check.
+
+  Add the ability to strictly enforce variables are used for specific classes,
+  even with they would not be normally warned using -Wunused-variable due 
```
Please highlight `-Wunused-variable` with single back-ticks.

rockwotj replied:
Thanks, done.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/misc/must-use.rst:
```diff
@@ -0,0 +1,28 @@
+misc-must-use
+=============
+
+Allows for strictly enforce variables are used for specific classes, even with
```
Please make this statement same as in Release Notes.

rockwotj replied:
Done.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/misc/must-use.rst:
```diff
@@ -0,0 +1,28 @@
+they would not be normally warned using -Wunused-variable due to templates or
+custom destructors.
+
+In the following example, future2 normally would not trigger any unused variable checks,
```
Please highlight `future2` with double back-ticks. Same for other occurrence.

rockwotj replied:
Thanks done.

PiotrZSL commented on clang-tools-extra/clang-tidy/misc/MiscTidyModule.cpp:
```diff
@@ -54,6 +55,7 @@ class MiscModule : public ClangTidyModule {
     CheckFactories.registerCheck<MisleadingIdentifierCheck>(
         "misc-misleading-identifier");
     CheckFactories.registerCheck<MisplacedConstCheck>("misc-misplaced-const");
+    CheckFactories.registerCheck<MustUseCheck>("misc-must-use");
```
I do not like check name, it's to short, and doesn't say what is not used (member, var, global, function, class ?).
Second this should eb a bugprone check with preconfigured things like std::mutex, std::future, and probably few other.
Third i personally got check called bugprone-unused-local-non-trivial-variable, that is implemented like this:
```
void BugproneUnusedLocalNonTrivialVariableCheck::registerMatchers(MatchFinder *Finder) {
  Finder->addMatcher(varDecl(unless(isExpansionInSystemHeader()),
                             unless(isImplicit()),
                             unless(parmVarDecl()),
                             unless(decompositionDecl()),
                             unless(isExceptionVariable()),
                             hasLocalStorage(),
                             isDefinition(),
                             unless(hasType(namedDecl(matchesName(ExcludeTypeRegex)))),
                             unless(hasType(qualType(references(namedDecl(matchesName(ExcludeTypeRegex)))))),
                             decl().bind("var"),
                                 hasAncestor(functionDecl(unless(hasDescendant(declRefExpr(to(decl(equalsBoundNode("var"))))))))
                            ).bind("var"), this);
}

void BugproneUnusedLocalNonTrivialVariableCheck::check(const MatchFinder::MatchResult &Result) {
  const auto *MatchedDecl = Result.Nodes.getNodeAs<VarDecl>("var");

  auto type = MatchedDecl->getType();
  if (type.isTrivialType(*Result.Context) or type.isTriviallyCopyableType(*Result.Context)) return;

  diag(MatchedDecl->getLocation(), "unused local variable %0 of non trivial type %1, consider removing")
      << MatchedDecl << type;
}

It's not perfect, but what I'm thinking, is if this check and that shoudn't be a single check with inclusion/exclusion settings, so it could be configured to warn on all or just on some limited scope.
```

rockwotj replied:
Merging that check and mine STGM (as well as the naming of your check). Do you have source for your check you'd like to point to? Otherwise I'm happy to add in a lot of these.

rockwotj replied:
Is there a difference between isReferenced versus the following:

```
decl().bind("var"),         hasAncestor(functionDecl(unless(hasDescendant(declRefExpr(to(decl(equalsBoundNode("var"))))))))
```

Sorry still new to AST matchers

PiotrZSL replied:
I think isReferenced is newer, so stick to it. Fell free to merge those, check that i pointed out is old, so basicly the main thing that is missing in yuors is a validation of trivial type and support for inclusion/exclusion matchers.
and few tests, so just start with that. rename check and we go from this.

rockwotj replied:
SG added most of these checks, let me know if you think I should add any.

PiotrZSL commented on clang-tools-extra/clang-tidy/misc/MustUseCheck.cpp:
```diff
@@ -0,0 +1,49 @@
+      Types(utils::options::parseStringList(Options.get("Types", ""))) {}
+
+void MustUseCheck::registerMatchers(MatchFinder *Finder) {
+  if (Types.empty()) {
```
no need for extra {}

rockwotj replied:
Done.

PiotrZSL commented on clang-tools-extra/clang-tidy/misc/MustUseCheck.cpp:
```diff
@@ -0,0 +1,49 @@
+
+void MustUseCheck::check(const MatchFinder::MatchResult &Result) {
+  const auto *MatchedDecl = Result.Nodes.getNodeAs<VarDecl>("var");
+  diag(MatchedDecl->getLocation(), "variable %0 must be used") << MatchedDecl;
```
print type in this message

rockwotj replied:
Done.

PiotrZSL commented on clang-tools-extra/clang-tidy/misc/MustUseCheck.h:
```diff
@@ -0,0 +1,40 @@
+  void registerMatchers(ast_matchers::MatchFinder *Finder) override;
+  void check(const ast_matchers::MatchFinder::MatchResult &Result) override;
+  void storeOptions(ClangTidyOptions::OptionMap &Opts) override;
+
```
consider excluding implicit code with TK_IgnoreUnlessSpelledInSource

rockwotj replied:
Done.

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/misc/must-use.rst:
```diff
@@ -0,0 +1,28 @@
+templates or custom destructors.
+
+In the following example, `future2` normally would not trigger any unused variable checks,
+but using `{key: "misc-must-use.Types", value: "std::future"}` would cause `future2` to have
```
this configuration is in deprecated style, use new one misc-must-use.Types: 'std::future'

rockwotj replied:
Thanks TIL, fixed.

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/misc/must-use.cpp:
```diff
@@ -0,0 +1,48 @@
+    auto Num2 = PendingA.get();
+    auto Num3 = qux(Num);
+    return Num * Num3;
+}
```
add tests with:
- variable type wrapped into other type, like std::shared_ptr<async::Future>
- async::Future used as an static variable
- async::Future used as global variable
- async::Future used as an function parameter
- async::Future captured into lambda
- async::Future in template function, like Future<T> Pending;
- async::Future used as reference: async::Future&
- async::Future used as pointer: async::Future*

rockwotj replied:
Done.

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/misc/must-use.cpp:
```diff
@@ -0,0 +1,48 @@
+// RUN: %check_clang_tidy %s misc-must-use %t -- \
+// RUN:       -config="{CheckOptions: [{key: 'misc-must-use.Types', value: '::async::Future'}]}"
```
write in new format.

rockwotj replied:
Done.

PiotrZSL commented on clang-tools-extra/clang-tidy/misc/MustUseCheck.h:
```diff
@@ -0,0 +1,40 @@
+  void registerMatchers(ast_matchers::MatchFinder *Finder) override;
+  void check(const ast_matchers::MatchFinder::MatchResult &Result) override;
+  void storeOptions(ClangTidyOptions::OptionMap &Opts) override;
+
```
consider restricting check to C++, or at least disable it when C is used

rockwotj replied:
Done.

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/misc/must-use.rst:
```diff
@@ -0,0 +1,28 @@
+=============
+
+Allow strictly enforcing that variables are used for specific classes,
+even with they would not be normally warned using `-Wunused-variable` due 
```
consider changing check in a way, so it would not overlap with -Wunused-variable

rockwotj replied:
Done.

PiotrZSL reviewed (changes requested):
Simple check, not bad. My main blocker for this check is a name.

rockwotj committed:
clang-tidy/bugprone: introduce unused-local-non-trivial-variable check

Signed-off-by: Tyler Rockwood <rockwood@redpanda.com>

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -168,6 +168,19 @@ New checks
+- New :doc:`bugprone-unused-local-non-trivial-variable
+  <clang-tidy/checks/bugprone/unused-local-non-trivial-variable>` check.
+
+  Warns when a local non trivial variable is unused within a function. By
```
Release Notes should contain single statement.

rockwotj replied:
Done.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unused-local-non-trivial-variable.rst:
```diff
@@ -0,0 +1,29 @@
+.. option:: IncludeTypeRegex
+
+   Semicolon-separated list of regular expressions matching types of variables to check.
+   By default it 'std::.*mutex;std::future'.
```
Please use single back-ticks for option names and values.

rockwotj replied:
Done.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unused-local-non-trivial-variable.rst:
```diff
@@ -0,0 +1,29 @@
+.. option:: ExcludeTypeRegex
+
+   A semicolon-separated list of regular expressions matching types that are excluded from the
+  'IncludeTypeRegex' matches. By default it is an empty list.
```
Ditto.

rockwotj replied:
Done.

dotnwat commented:
This is a very nice check for us to have @rockwotj.

PiotrZSL reviewed (changes requested):
Minor tuning needed, some changes to documentation, options, default configuration.

rockwotj commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/unused-local-non-trivial-variable.cpp:
```diff
@@ -0,0 +1,79 @@
+    a::Future<Units>& Reference = Referenced;
+    const a::Future<Units>& ConstReference = Referenced;
+    return Num * Num3;
+}
```
Done.

rockwotj commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/unused-local-non-trivial-variable.cpp:
```diff
@@ -0,0 +1,79 @@
+    a::Future<Units>& Reference = Referenced;
+    const a::Future<Units>& ConstReference = Referenced;
+    return Num * Num3;
+}
```
Done.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unused-local-non-trivial-variable.rst:
```diff
@@ -0,0 +1,45 @@
+* static or thread local
+* structured bindings
+
+This check can be configured to warn on all non-trivial variables by setting `ExcludeTypes` to `.*`.
```
Please follow 80-characters limit. Same in other places in this file.

rockwotj replied:
Fixed thanks.

rockwotj committed:
clang-tidy/bugprone: match dependent types in
unused-local-non-trivial-variable

Also add the ability to match against template specializations, this can
be useful if you want all std::future types to be matched, no matter
what type is held within the future.

Signed-off-by: Tyler Rockwood <rockwood@redpanda.com>

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/UnusedLocalNonTrivialVariableCheck.cpp:
```diff
@@ -0,0 +1,89 @@
+namespace clang::tidy::bugprone {
+
+namespace {
+static constexpr StringRef DefaultIncludeTypeRegex = "std::.*mutex;std::future";
```
I think that this std::string wont be detected because this is just typedef, and oryginal type is actually a std::base_string.

Consider adding those as default:
```
std::basic_istringstream
std::basic_stringstream
std::basic_string_view
std::basic_string
std::bitset
std::basic_regex
std::path
```

Those do not cary object ownership, when things like std::vector could.
I assume that anyway almost eveyrone will use configuration that catch all and use excludes. You could put into documentation that default check configuration is limited and recommended is to set allowed to .* to catch as much as possible.

rockwotj replied:
Done. I didn't add `std::basic_string_view` because string_view should be a trivial type.

PiotrZSL reviewed (approved):
Overall looks fine.

rockwotj commented on clang-tools-extra/clang-tidy/bugprone/UnusedLocalNonTrivialVariableCheck.cpp:
```diff
@@ -0,0 +1,92 @@
+static constexpr StringRef DefaultIncludeTypeRegex =
+    "::std::.*mutex;::std::future;::std::basic_string;::std::basic_regex;"
+    "::std::base_istringstream;::std::base_stringstream;::std::bitset;"
+    "::std::path";
```
Yeah I don't use these APIs much, but good catch! I've fixed it.

rockwotj commented on clang-tools-extra/clang-tidy/bugprone/UnusedLocalNonTrivialVariableCheck.cpp:
```diff
@@ -0,0 +1,92 @@
+
+  Finder->addMatcher(
+      varDecl(isLocalVarDecl(), unless(isReferenced()),
+              unless(isExpansionInSystemHeader()),
```
SG removed.

rockwotj commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/unused-local-non-trivial-variable.cpp:
```diff
@@ -0,0 +1,91 @@
+// RUN: %check_clang_tidy -std=c++17 %s bugprone-unused-local-non-trivial-variable %t -- \
```
Done.

rockwotj commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/unused-local-non-trivial-variable.cpp:
```diff
@@ -0,0 +1,91 @@
+// RUN: %check_clang_tidy -std=c++17 %s bugprone-unused-local-non-trivial-variable %t -- \
+// RUN:       -config="{CheckOptions: {bugprone-unused-local-non-trivial-variable.IncludeTypes: '::async::Future'}}"
```
Done.

PiotrZSL reviewed (approved):
LGTM.
Unless you want this to be merged by rockwotj@users.noreply.github.com please change your github privacy settings.

rockwotj commented:
Changed
